### ISSUE TITLE
Add shorthand reference field when self-references are required

### ISSUE THREAD

**Issue by headconnect** [pulumi/pulumi-azure-native#631]
## Affected feature
Quite a lot of azure.networking defenitions require subresources to reference other subresources of the resource one is trying to create. In my instance, ApplicationGateway resource requires the defenition of `requestRoutingRules`, which contain multiple references to other subresources of the ApplicationGateway, among which is the `httpListener`, which in turn requires references to `frontendIPConfigurations` and `frontendPorts`. 

Simply put - for resource A, subresource A.x references A.y which references A.v and A.w.  

While this is an issue with the underlying ARM templates themselves, the feature I would request to resolve it would be the introduction of a `name` (or `nameRef` or `subresourceName`) field which is outside the specification, xor with the `id` field, and would internally generate the correct `id:` for use in the arm template.

Take the following example specification of an ApplicationGateway: (apologies for it being huge- but that's part of the point I'm trying to make as well..)
```ts
const appGW = new azure-native.network.ApplicationGateway("myGW", {
    applicationGatewayName: config.require("appGWName"),
    autoscaleConfiguration: {
        maxCapacity: 20,
        minCapacity: 0
    },
    backendAddressPools: [{
        backendAddresses: [],
        name: "defaultbackendpool" // referenced by requestRoutingRules
    }],
    backendHttpSettingsCollection: [{
        name: "defaulthttpsetting", // referenced by requestRoutingRules
        requestTimeout: 60,
        port: 80,
    }],
    enableHttp2: config.requireBoolean("http2binary"),
    frontendIPConfigurations: [{
        name: "publicIp", // referenced by httpListeners
        publicIPAddress: {
            id: pubIp.id // reference to previously created azure-native.network.PublicIPAdress resource
        }
    }],
    frontendPorts: [{
        name: "httpport", // referenced by httpListeners
        port: 80
    }],
    gatewayIPConfigurations: [{}], // omitted for brevity,
    httpListeners: [{
        frontendIPConfiguration: {
            // This is where the problem lies - the id field is resolvable to the id of the resource we are currently generating, but can be "easily" calculated from resources we are already aware of.
            // the improvement would be to introduce a new field, as shown in the example below
            id: myResourceGroup.id.apply(id => `${id}/providers/Microsoft.Network/applicationGateways/${config.require("appGWName")}/frontendIPConfigurations/publicIp`),
        },
        frontendPort: {
            id: myResourceGroup.id.apply(id => `${id}/providers/Microsoft.Network/applicationGateways/${config.require("appGWName")}/frontendPorts/httpport`),
            
        },
        name: "defaulthttplistener", // referenced by requestRoutingRules
        protocol: "Http",
    }],
    location: config.require("location"),
    probes: [],
    requestRoutingRules: [{
        backendAddressPool: {
            id: myResourceGroup.id.apply(id => `${id}/providers/Microsoft.Network/applicationGateways/${config.require("appGWName")}/backendAddressPools/defaultbackendpool`),
        },
        backendHttpSettings: {
            id: myResourceGroup.id.apply(id => `${id}/providers/Microsoft.Network/applicationGateways/${config.require("appGWName")}/backendHttpSettingsCollection/defaulthttpsetting`),
        },
        httpListener: {
            id: myResourceGroup.id.apply(id => `${id}/providers/Microsoft.Network/applicationGateways/${config.require("appGWName")}/httpListeners/defaulthttplistener`),
        },
        name: "appgwrule"
    }],
    resourceGroupName: config.require("rgName"),
    sku: {
        name: "WAF_v2"
        tier: "WAF_v2"
    },
    sslCertificates: [],
    tags: config.requireObject("standardTags"),
    webApplicationFirewallConfiguration: {
        enabled: true,
        firewallMode: "Prevention",
        ruleSetType: "OWASP",
        ruleSetVersion: "3.0"
    },
```

An improved version of this where the subresource references do not need to manually caluculate IDs of subresources that do not yet exist would be like such:

```ts
const appGW = new azure-native.network.ApplicationGateway("myGW", {
    applicationGatewayName: config.require("appGWName"),
    autoscaleConfiguration: {
        maxCapacity: 20,
        minCapacity: 0
    },
    backendAddressPools: [{
        backendAddresses: [],
        name: "defaultbackendpool" // refrenced by requestRoutingRules
    }],
    backendHttpSettingsCollection: [{
        name: "defaulthttpsetting", // referenced by requestRoutingRules
        requestTimeout: 60,
        port: 80,
    }],
    enableHttp2: config.requireBoolean("http2binary"),
    frontendIPConfigurations: [{
        name: "publicIp", // referenced by httpListeners
        publicIPAddress: {
            id: pubIp.id // reference to previously created azure-native.network.PublicIPAdress resource
        }
    }],
    frontendPorts: [{
        name: "httpport", // referenced by httpListeners
        port: 80
    }],
    gatewayIPConfigurations: [{}], // omitted for brevity,
    httpListeners: [{
        frontendIPConfiguration: {
            // This is much better. We could simplify more by not even requireing the "#¤%"# field to be an object, but perhaps note it simply as
            // frontendIPConfigurationName: "publicIp"
            subResourceName: "publicIp"
        },
        frontendPort: {
            subResourceName: "httpport",
        },
        name: "defaulthttplistener", // referenced by requestRoutingRules
        protocol: "Http",
    }],
    location: config.require("location"),
    probes: [],
    requestRoutingRules: [{
        backendAddressPool: {
            subResourceName: "defaultbackendpool"
        },
        backendHttpSettings: {
            subResourceName: "defaulthttpsetting"
        },
        httpListener: {
            subResourceName: "defaulthttplistener"
        },
        name: "appgwrule"
    }],
    resourceGroupName: config.require("rgName"),
    sku: {
        name: "WAF_v2"
        tier: "WAF_v2"
    },
    sslCertificates: [],
    tags: config.requireObject("standardTags"),
    webApplicationFirewallConfiguration: {
        enabled: true,
        firewallMode: "Prevention",
        ruleSetType: "OWASP",
        ruleSetVersion: "3.0"
    },
```

Now I don't personally mind how you do it, but the less I have to calculate by hand (enumerate with string concatenation et cetera) the better. 

**TL:DR;** - Whenever there is an Input requiring an `id` of a SubResource type, create a "helper" field `subResourceName` that allows us to reference these with reusable strings.  Bonus points to eliminate these objects entirely and operate purely on subresource name references (as the `id` resolves to internally anyway).

**Comment by epomatti** [comment]
Excellent explanation. I didn't want to accept it had to be set explicitly, but here we are.

Perhaps docs should emphasize that upfront, would have saved me an hour or so.

I was looking for a shorthand and this `myResourceGroup.id.apply` helps.

**Comment by sunilshahi** [comment]
Wow, I wasted my whole day trying to figure out how to create an Application Gateway(Azure Native) and was disappointed when I ran into this issue here. I feel stupid after reading this. 

Please consider this change.

**Comment by mikhailshilkov** [comment]
Hi @headconnect and others, thank you for the detailed issue and all the upvotes, we realize that it's a common friction point and are happy to discuss a potential improvement.

I took a look at the problem and prototype a potential solution to the issue. I want to share my finding to seek for the community feedback.

First, I explored your proposal above, i.e. a new name-like property like

```
frontendIPConfiguration: {
    subResourceName: "publicIp"
},
```

The upsides are:

1. Really convenient for end users.
2. Similar to Azure Classic experience.

The downsides:

1. This requires a custom implementation on our side for each property. We would need to know which subresource type is implicitly referenced from each SubResource definition and resolve it accordingly. This is somewhat costly on the implementation side and error-prone if we miss a property or mistype the reference. It goes against the provider philosophy to be driven by API definitions and not manual implementation.
2. We reserve a property name which, in theory, can be introduced by Microsoft in the future. They may decide to solve the same issue on the API level in an incompatible way, which will bring us some trouble.

I have an alternative proposal that requires no changes to the SDK shape but will solve all the issues above. It's a bit more verbose for the user, but also more explicit.

### Short-hand ID syntax

The `id` property remains and can still accept the same values as currently. In addition, the subresource's `id` property will accept a _partial_ ID that is _relative to the parent resource_. For the example above, it would be

```ts
const appGW = new azure-native.network.ApplicationGateway("myGW", {
    // Note: no need for an explicit name anymore, auto-naming works
    backendAddressPools: [{
        backendAddresses: [],
        name: "defaultbackendpool" // referenced by requestRoutingRules
    }],
    backendHttpSettingsCollection: [{
        name: "defaulthttpsetting", // referenced by requestRoutingRules
        requestTimeout: 60,
        port: 80,
    }],
    frontendIPConfigurations: [{
        name: "publicIp", // referenced by httpListeners
        publicIPAddress: {
            id: pubIp.id // reference to previously created azure-native.network.PublicIPAdress resource
        }
    }],
    frontendPorts: [{
        name: "httpport", // referenced by httpListeners
        port: 80
    }],
    httpListeners: [{
        frontendIPConfiguration: {
            // Note that there is no apply here. No '/' in the beginning means that ID is relative.
            id: "frontendIPConfigurations/publicIp",
        },
        frontendPort: {
            id: "frontendPorts/httpport",
            
        },
        name: "defaulthttplistener", // referenced by requestRoutingRules
        protocol: "Http",
    }],
    requestRoutingRules: [{
        backendAddressPool: {
            id: "backendAddressPools/defaultbackendpool",
        },
        backendHttpSettings: {
            id: "backendHttpSettingsCollection/defaulthttpsetting",
        },
        httpListener: {
            id: "httpListeners/defaulthttplistener",
        },
        name: "appgwrule"
    }],
```

Here is another example, a LoadBalancer in Python. It uses variables to define links more explicitly. Still, it requires no outputs, applies, or explicit building of full IDs, and works with auto-naming.

```py
lb_name = f"{project}-lb"
lb_backend_name = f"{lb_name}-backend"
lb_backend_id = "backendAddressPools/" + lb_backend_name

lb_frontend_name = f"{lb_name}-frontend"
lb_frontend_id = "frontendIPConfigurations/" + lb_frontend_name

lb_probe_name = f"{lb_name}-probe"
lb_probe_id = "probes/" + lb_probe_name

lb = network.LoadBalancer(
    lb_name,
    backend_address_pools=[
        network.BackendAddressPoolArgs(
            name=lb_backend_name,
        ),
    ],
    frontend_ip_configurations=[network.FrontendIPConfigurationArgs(
        name=lb_frontend_name,
        public_ip_address=network.PublicIPAddressArgs(
            id=lb_public_ip.id,
        ),
    )],
    load_balancing_rules=[network.LoadBalancingRuleArgs(
        backend_address_pool=network.SubResourceArgs(
            id=lb_backend_id,
        ),
        backend_port=80,
        disable_outbound_snat=True,
        enable_floating_ip=False,
        enable_tcp_reset=True,
        frontend_ip_configuration=network.SubResourceArgs(
            id=lb_frontend_id,
        ),
        frontend_port=80,
        idle_timeout_in_minutes=4,
        load_distribution="Default",
        name=f"{lb_name}-rule1",
        probe=network.SubResourceArgs(
            id=lb_probe_id,
        ),
        protocol="TCP",
    )],
    probes=[network.ProbeArgs(
        interval_in_seconds=5,
        name=lb_probe_name,
        number_of_probes=1,
        port=80,
        protocol="HTTP",
        request_path="/",
    )],
    resource_group_name=resource_group.name,
    sku=network.LoadBalancerSkuArgs(
        name="Standard",
    ),
)
```

I'm curious what folks think about this solution. I have a hacked-together version of it working locally, so it's definitely realistic to ship it.

**Comment by mikhailshilkov** [comment]
A draft PR is ready in #2478

A quick poll around which syntax you prefer.

Option 🎉  - relative URL, parent ID will be prepended

```
frontendIPConfiguration: {
    id: "frontendIPConfigurations/publicIp",
}
```

Option 🚀 - explicit placeholder that will be replaced by parent ID

```
frontendIPConfiguration: {
    id: "$self/frontendIPConfigurations/publicIp",
}
```

Feel free to comment with your considerations.

**Comment by headconnect** [comment]
So sorry for not responding to this earlier - I saw it, then started thinking, then got lost in work. Simply put - I'm at the stage where I'm just super happy to see that this is getting resolved, and all options presented here seem to do it in a good way.
